File: packages/shared/src/memory.ts
import { z } from "zod";

// --- Enums ---

export const MemoryScope = z.enum(["project", "agent", "archive"]);
export type MemoryScope = z.infer<typeof MemoryScope>;

export const MemoryType = z.enum([
  "knowledge",
  "decision",
  "lesson",
  "error",
  "log",
  "handoff_summary",
  "merge_summary",
]);
export type MemoryType = z.infer<typeof MemoryType>;

// --- Schema ---

export const MemorySchema = z.object({
  id: z.string().uuid(),
  agentId: z.string().uuid().nullable(),
  scope: MemoryScope,
  moduleId: z.string().uuid().nullable(),
  type: MemoryType,
  content: z.string(),
  sourceAgentId: z.string().uuid().nullable(),
  metadata: z.record(z.string(), z.unknown()),
  createdAt: z.string().datetime(),
});
export type Memory = z.infer<typeof MemorySchema>;

File: apps/hiveweave/lib/hiveweave/schema/memory.ex
defmodule HiveWeave.Schema.Memory do
  use Ecto.Schema
  import Ecto.Changeset

  schema "memories" do
    field :agent_id, :string
    field :scope, :string  # project | agent_private | archive
    field :module_id, :string
    field :type, :string
    field :content, :string
    field :source_agent_id, :string
    field :metadata, :string, default: "{}"
    field :created_at, :integer
    field :updated_at, :integer

  end

  def changeset(memory, attrs) do
    memory
    |> cast(attrs, [
      :agent_id, :scope, :module_id, :type, :content, :source_agent_id,
      :metadata, :created_at, :updated_at
    ])
    |> validate_required([:scope, :type, :content, :created_at, :updated_at])
  end
end
